### PR TITLE
ncm-nmstate: relax syntax of VLAN interface names

### LINKED ISSUE #1678: ncm-nmstate: VLAN name pattern too restrictive
`ncm-nmstate` changes in https://github.com/quattor/configuration-modules-core/issues/1667 recognizes as a VLAN any interface whose name starts with `vlan` but imposes a `.` before the VLAN ID, despite it is not an OS or `nmstate` requirement.

### PR DESCRIPTION
**Pull request by jouvin**
Fixes #1678

### CODE REVIEW

**Comment by aka7** [comment]
When I did this for backward compatibility for network.pm, I followed this test file.
https://github.com/quattor/configuration-modules-core/blob/master/ncm-network/src/test/resources/route_rule.pan#L36

so I think long as your new change still covers this? then it should be fine for backward compatibility with network.pm?

Due to legacy code we had internally, we only define vlan interface using interfacename.vlan_id, i.e eth0.123 format  as part of the interface name when defining vlan iface.  So long as this test passes,
https://github.com/quattor/configuration-modules-core/blob/master/ncm-network/src/test/resources/nmstate_advance.pan#L30
 we should be ok on our end.
so LGTM but I will test your PR against our profiles later this week to confirm.

regards

**Comment by jouvin** [comment]
To be clear, my PR currently only relax the pattern matching on interface/device name but doesn't implement my second suggestion of inverting the order in which we use interface name and device to guess the VLAN ID. As for us this second point is not very important because we always use the same value for the interface name and the device. But I have the feeling you would allow the interface name not to convey the VLAN ID but have it attached to the device name instead...

Anyway, with my fix, I'm a bit surprised that `vlan0` test is still working... I should look in more details. But this fix has been deployed at IJCLab and fixes the problem we had before as our VLAN interfaces are all named `vlanxxx`, `xxx` being the VLAN ID.

**Comment by aka7** [comment]
@jouvin from what I see, vlan0 test config is working as expected with your change since you just relaxed it and not removed anything. vlan0 config is using device=eth0.123 to get the vlanid.  (https://github.com/quattor/configuration-modules-core/blob/master/ncm-network/src/test/resources/nmstate_advance.pan#L40)  so that is correct and I wouldn't have expected your change to break it?  so IMO it is working as expected. Unless I'm misunderstanding the solution here.

I'm not fussed with order of check,  as we are only setting it one way which is using the format  "/system/network/interfaces/eth0.123"

so long as this format of config carries on working too, it should be fine.

I just did quick check on lab host and its a noop as expected.

**Comment by jouvin** [comment]
@aka7 good! I'll see if I can come up with the reverse order giving the current result. For me it should...

My point with vlan0 is that with the relaxed regex, the ifnzme should match the regex and the vlan ID should be set to 0 which is not what we want. And for me the fix would be to reverse the order as the device will be carry the actual vlan0 ID and will be tested first. I guess it is working because ID 0 is causing the test to evaluate to false. My initial attempt to add vlan1 was failing in fact because it returned ID 1 instead of 123.

**Comment by jouvin** [comment]
@aka7 @stdweird I'm trying to fix a problem I identified with the way returned VLAN ID from the interface/device name is tested. The fix is trivial but I'm trying to add a test for this and I'm having trouble because adding the needed lines in the test profile make all the tests failing (including the first one which is just compiling the profile if I'm right) with something that looks as a Pan error except that compiling the profile manually gives no error. Surely a trivial mistake of mine... Any idea?

**Comment by aka7** [comment]
@jouvin  there is planned meetup discussion today,  are you able to join todays call, perhaps we can discuss it there?

**Comment by jouvin** [comment]
@aka7 sorry, I'm at a conference this week, it was impossible for me to join the meetup... Don't spend time trying to fix tests, I have almost everything done but currently I only published the side issue I identified and I don't understand why adding a VLAN interface in the test profile beaks the Pan compilation despite Pan being able to compile it successfully if run manually...

**Comment by jouvin** [comment]
@stdweird any chance you could have a look why  the second commit is breaking all tests... I'm sure it's trivial and you'll identify the problem very quickly! Just my lack of Perl/Quattor development these last years!

**Comment by jouvin** [comment]
@stdweird I'm still looking for some help to sort out my (probably stupid) mistake with the test additions, so that I can push the real fix related to the issue...

**Comment by stdweird** [comment]
@jouvin sorry, i don't know where my head is last weeks ;)
can you already rebase the PR?

**Comment by jouvin** [comment]
@stdweird don't worry, I'm totally overwhelmed too these days, tackling to many issues a the same time... My question was not really on the code itself (the line you mentioned is not from me and I was not completely convinced if was correct with Perl, look at a bashism...) but anyway my main problem at this stage is that I added one test in one of the `.t` file and this breaks all tests. It has nothing to do with the module code itself: just removing the added lines in the test is enough to make all the other tests successful. I guess I made a very stupid mistake in my test additions.

As for your request to rebase, I'd prefer not to do it at this stage (as I have the whole fix ready for this version and would prefer to rebase afterwards), except if you have particular reason for asking for it...

**Comment by stdweird** [comment]
hmmm, that is not the reason. if you can rebase, i can run the tests locally and debug it quicker

**Comment by jouvin** [comment]
Sorry I missed the conflicts... I'll do the rebase then...

**Comment by jouvin** [comment]
@stdweird rebase done

**Comment by stdweird** [comment]
@jouvin the pan unittest is an actual example? what is the generated filename for that device? probably the generated yml filename also needs to follow the `$device` logic you modified.

**Comment by jouvin** [comment]
@stdweird yes sure, I need to fix the test for the modified logic but wanted to fix the test logic first... I apologise missing the error, I thought it looked every line... My bad! I'll check in the next days and try to complete this PR.

**Comment by stdweird** [comment]
if you want to allow such names, just replace the whole device regex with `\w+`; or implement logic to skip the test (maybe controlled by boolean from the schema. you can have it default to true (ie allow arbitrary names by default)

**Comment by jouvin** [comment]
@stdweird yes it is what I have done in fact but I have not pushed it yet.. I wanted to add a test first demonstrated the problem with the `vlan0` example which works because of a bug in the configuration module...

For my information, where have you seen this "hidden error"?

**Comment by stdweird** [comment]
it's not hidden, it's just a single line of ERROR output; it's easily missed

**Comment by jouvin** [comment]
@stdweird it's more and more weird! I cannot see this line, even using `grep`. Here what I did:

```bash
$ mvn clean test -Dunittest=nmstate_advance.t >& /tmp/log.relax
$ grep -i ERROR /tmp/log.relax
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.4.0:exec (prove) on project network: Command execution failed. Process exited with an error: 1 (Exit value: 1) -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

I am not sure anyway what the missing file error you reported means... I've the feeling that the file should be created automatically from the `pan` resource provided, isn't it?

**Comment by jouvin** [comment]
I think I spotted the reason of the problem, but not sure why/where to fix it. It seems that the component doesn't like the device name `vlan.0` because there is no number after `vlan` and before the `.`. If using `vlan0.0`, the problem disappears. The change made seems to cause a problem with the `ib1.12345` test but it is something else and doesn't break all the other tests that receive `undef` instead of the expected yml file contents...

@aka7 @stdweird any idea why this happens? I'll look at the component code creating the file name, I guess it is parsing the interface name in some way...

**Comment by jouvin** [comment]
And it is unexpected for me. This error comes from `ncm-network` and I thought it was not used at all when using `ncm-nmstate`. I may have missed something...

**Comment by stdweird** [comment]
https://github.com/quattor/configuration-modules-core/blob/master/ncm-network/src/main/perl/network.pm#L156 is the regex.

**Comment by jouvin** [comment]
@stdweird I found it but is it expected that we execute some ncm-network code when using ncm-nmstate?

**Comment by stdweird** [comment]
@jouvin yes, that is normal. the nmstate backend is a subclass of the network backend

**Comment by jouvin** [comment]
Hi, I have been unable to finish the work on this critical fix for us (required if your VLAN name is not containing a `.` after `vlan` prefix, which is the case at IJCLab. The fix is working well but unfortunately I have been struggling with the tests failing for an unidentified reason so far... I rebased but it doesn't help (I was not really expecting it to help)... Not sure about what to do...

**Comment by stdweird** [comment]
@jouvin the tests still fail on the same error. you need to adapt the regex to pass them

**Comment by jouvin** [comment]
Thanks @stdweird . I forgot about it when doing my tests yesterday... Modifying this regex is really worrying... but I'll give it a try, hopefully this week :-)

**Comment by jouvin** [comment]
I add a look this morning. The problem is not `DEVICE_REGEX` as the `is_valid_interface()` method is redefined in `nmstate` and is using another regex! I'm progressing!

**Comment by jouvin** [comment]
I was able to look in details into the remaining problems and to the things already fixed. It is mainly the problem of the filename regexp that I relaxed in my last changes. I added a test for `vlan.456` but I have a failure during the `apply` test. And again I have no clue about what the problem could be, despite `-Dprove.args=-v`. @stdweird any suggestion where to look? In particular, is there a way to look at the generated YML file contents?

**Comment by jouvin** [comment]
Answering to myself... I found the error: `command_history_ok` check the command order and it was not what was expected... I should have read the command saying the commands are sorted alphabetically, `bond` last... I guess this sorting is done by `nmstate` command...

Anyway, the first for the original issue is now complete and working. @jrha is there any possibility to have it included inside the release? It is really critical for us as all our VLANs are name `vlan.nnn` and we don't want to change that.

Note: the changes include a fix for `vlan0` expected config which IMO was wrong due to a bug fixed in 4dc5fff13bfecd9bcdbb9250aa19a6b657b94804.

File: ncm-network/src/main/perl/nmstate.pm [hunk 1]
```diff
@@ -64,8 +66,8 @@ sub is_valid_interface
             # Filename is either right at the beginning or following a slash
             (?: \A | / )
             # $1 will capture for example:
-            # eth0  bond1  eth0.101  bond0.102 or dummy_$key
-            ( \w+ \d+ (?: \. \d+ )? | dummy_.* )
+            # eth0  bond1  eth0.101  bond0.102 vlan.456 or dummy_$key
+            ( \w+ \d+ (?: \. \d+ )? | \w+ \. \d+ | dummy_.* )
```
[the remark this answers is not in the document]
Comment (jouvin) [reply ncm-network/src/main/perl/nmstate.pm:70]: @jrha I'm happy with your suggestion, it was my first implementation in fact... But as Linux doesn't require a VLAN interface to start with `vlan` if I'm right, I thought it was probably too restrictive and may cause problem if somebody wants to have another naming scheme... So I decided to go with this more open regex, still requiring either something like `eth0[.123]` or `xyz.123` as one of the tests in `nmstate_simple` checks that anything else is rejected. But let me know if you prefer the version you suggested, it's fine with me.

**Review by jrha — DISMISSED** [review]

**Comment by jrha** [comment]
Assuming @aka7 has no problems with this, I'll get it merged tomorrow and start on another RC.

**Comment by jrha** [comment]
Rebased and conflicts addressed.

**Review by jrha — APPROVED** [review]

**Pull request merged**